src/artemis/external_interaction/callbacks/fgs/zocalo_callback.py

```diff
@@ -42,30 +44,39 @@ def __init__(
         ] = parameters.grid_scan_params.grid_position_to_motor_position
         self.processing_start_time = 0.0
         self.processing_time = 0.0
+        self.results = None
+        self.started_run: bool = False
+        self.run_gridscan_uid: Optional[str] = None
+        self.xray_centre_motor_position = None
         self.ispyb = ispyb_handler
         self.zocalo_interactor = ZocaloInteractor(parameters.zocalo_environment)
 
-    def event(self, doc: dict):
-        LOGGER.debug("Zocalo handler received event document.")
-        descriptor = self.ispyb.descriptors.get(doc["descriptor"])
-        assert descriptor is not None
-        event_name = descriptor.get("name")
-        if event_name == ISPYB_PLAN_NAME:
+    def start(self, doc: dict):
+        LOGGER.info("Zocalo handler received start document.")
+        if doc.get("subplan_name") == "do_fgs":
+            self.run_gridscan_uid = doc.get("uid")
             if self.ispyb.ispyb_ids[0] is not None:
                 datacollection_ids = self.ispyb.ispyb_ids[0]
                 for id in datacollection_ids:
                     self.zocalo_interactor.run_start(id)
+                self.started_run = True
             else:
                 raise ISPyBDepositionNotMade("ISPyB deposition was not initialised!")
 
     def stop(self, doc: dict):
-        LOGGER.debug("Zocalo handler received stop document.")
-        if self.ispyb.ispyb_ids == (None, None, None):
-            raise ISPyBDepositionNotMade("ISPyB deposition was not initialised!")
-        datacollection_ids = self.ispyb.ispyb_ids[0]
-        for id in datacollection_ids:
-            self.zocalo_interactor.run_end(id)
-        self.processing_start_time = time.time()
+        if self.started_run:
+            if doc.get("run_start") == self.run_gridscan_uid:
+                LOGGER.info(
+                    f"Zocalo handler received stop document, for run {doc.get('run_start')}, and started run = {self.started_run}, uid : {self.run_gridscan_uid}"
```

Comment:
no, there are many sub-runs ("run_gridscan", "do_fgs", "move_xyz" ...; set with `@bpp.set_run_key_decorator()`) and we care about the stop document associated with the run_start uid for the sub-plan which the zocalo handler is using for a trigger

this will be tidier after #468 but that is probably longer-term

Reply:
oh, whoops, I misunderstood your comment at first glance. agree.